### PR TITLE
Keys seperate allocations

### PR DESCRIPTION
**Pull request by yonigottesman**
allocate keys directly from ByeBuffer.allocateDirect() without going through the nativeallocator.
This way we can remove start() stop() operations because keys will be release by ByteBuffer framwork when the reference count is 0.

This is needed because druid keeps a reference to the key bytebuffer while iterating and it must not be deleted.

### CODE REVIEW

File: core/src/main/java/com/oath/oak/InternalOakMap.java [hunk 1]
```diff
@@ -796,14 +768,9 @@ public int getIndex() {
...
 
-        boolean validState() { return state == State.VALID_STATE;}
 
-        public static <K, V> IteratorState<K, V> endState() {return endState;}
-        public static <K, V> IteratorState<K, V> state(Chunk<K, V> nextChunk, Chunk.ChunkIter nextChunkIter, int nextIndex) {
-            return new IteratorState<>(nextChunk, nextChunkIter, nextIndex, State.VALID_STATE);
-        }
         public static <K, V> IteratorState<K, V> initState(Chunk<K, V> nextChunk, Chunk.ChunkIter nextChunkIter) {
-            return new IteratorState<>(nextChunk, nextChunkIter, Chunk.NONE, State.INIT_STATE);
+            return new IteratorState<>(nextChunk, nextChunkIter, Chunk.NONE);
```
Comment (eranmeir) [inline core/src/main/java/com/oath/oak/InternalOakMap.java]: Perhaps capitalize this method or rename it to emphasize this is a static factory method and not an ordinary method. Static factory methods [often](https://docs.oracle.com/javase/tutorial/datetime/overview/naming.html) start with `from` or `of`.

File: core/src/main/java/com/oath/oak/InternalOakMap.java [hunk 2]
```diff
@@ -927,90 +886,57 @@ private boolean updateEpoch() {
...
-            //Init state, not valid yet, must move forword
-            IteratorState<K, V> initState = IteratorState.initState(nextChunk, nextChunkIter);
-            state.set(initState);
-
-            lastKeyBeforeRelease = new AtomicReference<>(initState.getChunk());
-
-            // ok to signin init state, the rebalancer will ignore
-            initState.getChunk().signInIterator(this);
 
+            //Init state, not valid yet, must move forword
```
Comment (eranmeir) [inline core/src/main/java/com/oath/oak/InternalOakMap.java]:
```suggestion
            //Init state, not valid yet, must move forward
```

File: core/src/main/java/com/oath/oak/MemoryManager.java [hunk 3]
```diff
@@ -6,147 +6,45 @@
...
-            this.timeStamps[i] = new AtomicLong(0);
-        }
-        this.releasedArray = new ArrayList<>(ThreadIndexCalculator.MAX_THREADS);
-        for (int i = 0; i < ThreadIndexCalculator.MAX_THREADS; i++) {
-            releasedArray.add(i, new LinkedList<>());
-        }
-        max = new AtomicLong(0);
-        releases = RELEASES_DEFAULT;
-        this.threadIndexCalculator = threadIndexCalculator;
+        keysMemoryAllocator = new DirectMemoryAllocator();
```
Comment (eranmeir) [inline core/src/main/java/com/oath/oak/MemoryManager.java]: I think concerns are separated better if you inject `keysMemoryAllocator` as a constructor argument (and allocate a new `DirectMemoryAllocator` if null)

File: core/src/test/java/com/oath/oak/OakMemoryManagerTest.java [hunk 4]
```diff
@@ -53,325 +53,4 @@ public void allocate() {
         assertEquals(4, bb.remaining());
         assertEquals(8, memoryManager.allocated());
     }
```
Comment (eranmeir) [inline core/src/test/java/com/oath/oak/OakMemoryManagerTest.java]: Why did you remove unit tests?

File: core/src/test/java/com/oath/oak/IteratorModificationTest.java [hunk 5]
```diff
@@ -231,87 +233,55 @@ public void concurrentModificationTest() throws InterruptedException {
...
-//    }
+
+    @Test
+    public void valueDeleteTest() throws InterruptedException {
+
+        OakIterator<Map.Entry<String, String>> enrtyIterator = oak.entriesIterator();
+        assertTrue(enrtyIterator.hasNext());
+        oak.remove(generateString(0, KEY_SIZE));
+        assertTrue(enrtyIterator.hasNext());
+        assertEquals(null, enrtyIterator.next());
```
Comment (eranmeir) [inline core/src/test/java/com/oath/oak/IteratorModificationTest.java]:
```suggestion
        assertNull(enrtyIterator.next());
```

File: core/src/test/java/com/oath/oak/OakMemoryManagerTest.java (hunk 4 above)
Comment (yonigottesman) [reply core/src/test/java/com/oath/oak/OakMemoryManagerTest.java]: because they dont mean anything anymore. they check when the release list is freed but in this new implementation there is no release list
